src/System.Management.Automation/engine/remoting/common/RemotingUtils.cs

```diff
@@ -0,0 +1,118 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using Microsoft.PowerShell.Commands;
+
+namespace System.Management.Automation.Remoting
+{
+    /// <summary>
+    /// Defines generic utilities and helper methods for Remoting.
+    /// </summary>
+    internal static class RemotingUtils
+    {
+        private const string s_LOCALHOST = "localhost";
+
+        /// <summary>
+        /// Parse a hostname used with SSH Transport to get embedded
+        /// username and/or port.
+        /// </summary>
+        /// <param name="cmdlet">PSRemoting base cmdlet.</param>
+        /// <param name="hostname">Host name to parse.</param>
+        /// <param name="host">Resolved target host.</param>
+        /// <param name="userName">Resolved target user name.</param>
+        /// <param name="port">Resolved target port.</param>
+        internal static void ParseSshHostName(PSRemotingBaseCmdlet cmdlet, string hostname, out string host, out string userName, out int port)
+        {
+            host = hostname;
+            userName = cmdlet.UserName;
+            port = cmdlet.Port;
+            try
+            {
+                Uri uri = new("ssh://" + hostname);
+
+                if (uri.HostNameType == UriHostNameType.Dns)
+                {
+                    // Extract original host from URI with preserved case
+                    // This is needed since System.Uri canonicalizes URI and makes host lowercase
+                    int originalHostIndex = uri.OriginalString
+                        .AsSpan()
+                        .IndexOf(uri.Host.AsSpan(), StringComparison.OrdinalIgnoreCase);
+
+                    string originalHost = originalHostIndex != -1
+                        ? new string(uri.OriginalString
+                            .AsSpan()
```

Comment:
Just as an FYI using a Span here and above vs your original Substring probably isn't doing anything more efficient. It is still going to allocate a new string just like `Substring` is going to do.

Reply:
So as expected not much difference:

```markdown
| Method      | NumberOfHostnames | Mean     | Error     | StdDev    |
|------------ |------------------ |---------:|----------:|----------:|
| WithSpan    | 10000             | 3.089 ms | 0.0588 ms | 0.0550 ms |
| WithoutSpan | 10000             | 3.112 ms | 0.0516 ms | 0.0457 ms |
```

Slightly lower mean execution time for Span and slightly less memory allocated for Span as well. The Gen0 allocations is the same so garbage collection is generating the same for both approaches. Overall it seems there is a small perf gain but probably not really worth to include since SubString is performing as good like you said :).

Reply:
Reverted. Benchmark shows it's not really worth the perf optimization.